src/container/LinkedCellContainer.cpp

```diff
@@ -0,0 +1,390 @@
+    return index3D[0] < 0 || index3D[0] >= nCells[0] || index3D[1] < 0 || index3D[1] >= nCells[1] || index3D[2] < 0 || index3D[2] >= nCells[2];
+}
+
+std::vector<int> LinkedCellContainer::getAllHaloIndices(Direction direction) {
+    std::vector<int> haloIndices;
+    switch (direction) {
+        case LEFT:
+            for (int z = -1; z <= nCells[2] ; ++z) {
+                for (int y = -1; y <= nCells[1]; ++y) {
+                    haloIndices.push_back(get1DIndex({-1, y, z}));
+                }
+            }
+            break;
+        case RIGHT:
+            for (int z = -1; z <= nCells[2] ; ++z) {
+                for (int y = -1; y <= nCells[1]; ++y) {
+                    haloIndices.push_back(get1DIndex({nCells[0], y, z}));
+                }
+            }
+            break;
+        case DOWN:
+            for (int z = -1; z <= nCells[2] ; ++z) {
+                for (int x = -1; x <= nCells[0]; ++x) {
+                    haloIndices.push_back(get1DIndex({x, -1, z}));
+                }
+            }
+            break;
+        case UP:
+            for (int z = -1; z <= nCells[2] ; ++z) {
+                for (int x = -1; x <= nCells[0]; ++x) {
+                    haloIndices.push_back(get1DIndex({x, nCells[1], z}));
+                }
+            }
+            break;
+        case BACK:
+            for (int y = -1; y <= nCells[1] ; ++y) {
+                for (int x = -1; x <= nCells[0]; ++x) {
+                    haloIndices.push_back(get1DIndex({x, y, -1}));
+                }
+            }
+            break;
+        case FRONT:
+            for (int y = -1; y <= nCells[1] ; ++y) {
+                for (int x = -1; x <= nCells[0]; ++x) {
+                    haloIndices.push_back(get1DIndex({x, y, nCells[2]}));
+                }
+            }
+    }
+    return haloIndices;
+}
+
+void LinkedCellContainer::removeFromHalo(Direction direction) {
+    for(int i : getAllHaloIndices(direction)) {
+        for (int p: cells[i].getParticleIndices()) {
+            (*particles)[p].removeFromDomain();
+            cells[i].removeIndex(p);
+        }
+    }
+}
+
+void LinkedCellContainer::updateHalo(Direction direction, BoundaryCondition boundaryCondition) {
+    for(int i : getAllHaloIndices(direction)) {
+        for (int p : cells[i].getParticleIndices()) {
+            switch (boundaryCondition) {
+                case OUTFLOW:
+                    (*particles)[p].removeFromDomain();
+                    cells[i].removeIndex(p);
+                    break;
+                case REFLECTING:
+                    // change sign of the corresponding velocity component if the particle is flying away.
+                    if ((*particles)[p].getV()[direction / 2] * (direction % 2 - 0.5) > 0) {
+                        std::array<double, 3> vel = (*particles)[p].getV();
+                        vel[direction / 2] *= -1;
+                        (*particles)[p].setV(vel);
+                    }
```

Comment:
We actually don't distinguish halo cells from normal cells in the force calculation, so particles in the halo cells are also considered (In case of outflow, particles are immediately removed from halo cells). By flipping the sign, the particle will most likely return to domain in the next time step. Ofcourse, there is potentially a problem if the force is too large as the update of x contains a force term. However, since this term scales as delta_t^2, choosing delta_t sufficiently small should avoid this problem. Nevertheless, I will consider making it more stable.